src/crystal/system/win32/dir.cr

```diff
@@ -130,12 +130,39 @@ module Crystal::System::Dir
   end
 
   def self.delete(path : String, *, raise_on_missing : Bool) : Bool
-    return true if LibC._wrmdir(System.to_wstr(path)) == 0
+    win_path = System.to_wstr(path)
 
-    if !raise_on_missing && Errno.value == Errno::ENOENT
-      false
+    attributes = LibC.GetFileAttributesW(win_path)
+    if attributes == LibC::INVALID_FILE_ATTRIBUTES
+      check_not_found_error("Unable to remove directory", path)
+      raise ::File::Error.from_os_error("Unable to remove directory", Errno::ENOENT, file: path) if raise_on_missing
+      return false
+    end
+
+    # all reparse point directories should be deleted like a directory, not just
+    # symbolic links, so we don't care about the reparse tag here
+    if attributes.bits_set?(LibC::FILE_ATTRIBUTE_REPARSE_POINT) && attributes.bits_set?(LibC::FILE_ATTRIBUTE_DIRECTORY)
+      # maintain consistency with POSIX, and treat all reparse points (including
+      # symbolic links) as non-directories
+      raise ::File::Error.new("Cannot remove directory that is a reparse point: '#{path.inspect_unquoted}'", file: path)
+    end
+
+    return true if LibC._wrmdir(win_path) == 0
+    raise ::File::Error.from_errno("Unable to remove directory", file: path)
+  end
+
+  NOT_FOUND_ERRORS = {
+    WinError::ERROR_FILE_NOT_FOUND,
+    WinError::ERROR_PATH_NOT_FOUND,
+    WinError::ERROR_INVALID_NAME,
+  }
+
+  private def self.check_not_found_error(message, path)
+    error = WinError.value
+    if NOT_FOUND_ERRORS.includes? error
+      nil
     else
-      raise ::File::Error.from_errno("Unable to remove directory", file: path)
+      raise ::File::Error.from_os_error(message, error, file: path)
     end
```

Comment:
This duplication seems unnecessary. Why not re-use `File.check_not_found_error` and call that from `.delete`?